// @lumiastream/plugin/dist/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plugin = void 0;
/**
 * Base Plugin Class - every Lumia Stream plugin should extend this class.
 */
class Plugin {
    constructor(_manifest, context) {
        this.context = context;
    }
    /**
     * Called when Lumia upgrades the plugin to a new version.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async onupdate(_oldVersion, _newVersion) { }
    /**
     * Handle custom actions triggered from the UI or Lumia automations.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async actions(_config) { }
    /**
     * Optional AI prompt handler used when `config.hasAI` is enabled.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async aiPrompt(_config) { }
    /**
     * Optional AI model list handler used when `config.hasAI` is enabled.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async aiModels(_config) {
        return [];
    }
    /**
     * Handle dashboard/API moderation commands when declared in manifest config.modcommandOptions.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async modCommand(_type, _value) { }
    /**
     * Refresh dynamic action field options in the UI.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async refreshActionOptions(_config) { }
    /**
     * Refresh dynamic settings field options in the PluginAuth UI.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async refreshSettingsOptions(_config) { }
    /**
     * Invoked whenever plugin settings are changed.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async onsettingsupdate(_settings, _previousSettings) { }
    get settings() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.lumia) === null || _b === void 0 ? void 0 : _b.getSettings) === null || _c === void 0 ? void 0 : _c.call(_b)) !== null && _d !== void 0 ? _d : {};
    }
    set settings(value) {
        var _a, _b;
        if ((_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.lumia) === null || _b === void 0 ? void 0 : _b.setSettings) {
            this.context.lumia.setSettings(value);
        }
    }
    updateSettings(updates) {
        var _a, _b;
        if ((_b = (_a = this.context) === null || _a === void 0 ? void 0 : _a.lumia) === null || _b === void 0 ? void 0 : _b.updateSettings) {
            this.context.lumia.updateSettings(updates);
        }
    }
    get manifest() {
        return this.context.plugin;
    }
    get lumia() {
        return this.context.lumia;
    }
    displayChat(options) {
        this.lumia.displayChat(options);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async validateAuth(_data) {
        return true;
    }
    async refreshAuth(data) {
        return data;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async searchLights(_query) {
        return [];
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async searchThemes(_query) {
        return [];
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async addLight(_data) {
        return null;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async searchPlugs(_query) {
        return [];
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async addPlug(_data) {
        return null;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async onLightChange(_config) {
        return;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async onPlugChange(_config) {
        return;
    }
}
exports.Plugin = Plugin;
exports.default = Plugin;
