// @lumiastream/plugin/dist/types.js
"use strict";
/**
 * Lumia Stream Plugin SDK - Core Types and Interfaces
 * Provides the type system that plugin developers rely on when building plugins.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCategory = exports.PluginStatus = void 0;
/**
 * Plugin lifecycle status as represented inside Lumia Stream.
 */
var PluginStatus;
(function (PluginStatus) {
    PluginStatus["INSTALLED"] = "installed";
    PluginStatus["ACTIVE"] = "active";
    PluginStatus["INACTIVE"] = "inactive";
    PluginStatus["ERROR"] = "error";
    PluginStatus["UPDATING"] = "updating";
    PluginStatus["PENDING_INSTALL"] = "pending_install";
    PluginStatus["UNINSTALLING"] = "uninstalling";
})(PluginStatus || (exports.PluginStatus = PluginStatus = {}));
/**
 * High level plugin marketplace categories.
 */
var PluginCategory;
(function (PluginCategory) {
    PluginCategory["SYSTEM"] = "system";
    PluginCategory["PLATFORMS"] = "platforms";
    PluginCategory["APPS"] = "apps";
    PluginCategory["AUDIO"] = "audio";
    PluginCategory["VIDEO"] = "video";
    PluginCategory["LIGHTS"] = "lights";
    PluginCategory["SWITCH"] = "switch";
    PluginCategory["DECK"] = "deck";
    PluginCategory["PROTOCOLS"] = "protocols";
    PluginCategory["KEYLIGHT"] = "keylight";
    PluginCategory["DEVICES"] = "devices";
    PluginCategory["GAMES"] = "games";
    PluginCategory["UTILITIES"] = "utilities";
})(PluginCategory || (exports.PluginCategory = PluginCategory = {}));
